### ISSUE TITLE
Questions on Loss Function Implementations

### ISSUE THREAD

**Issue by wenhaoqiao** [xionghonglin/Arbitrary_Contrast_MRI_Synthesis#1]
Hi author,

First, thank you for your excellent work on the MedIA paper and for making the source code publicly available. It's a very interesting and insightful project. I've been studying the paper and the codebase in detail and had a couple of questions regarding the implementation of the loss functions, where there seem to be some differences between the paper's description and the final code.

Q1. Content Alignment Loss
The paper (Section 3.2.1) describes using a memory bank to collect content representations from different forward passes within the same batch to compute the alignment loss. In the `train_brats.py`, however, loss_content seems to be calculated immediately in a forward pass, and I did not find the corresponding memory library code. If the current version is a simplified version, then will the content feature of the missing mode being 0 affect this loss calculation?

Q2. Orthogonality Loss
There appears to be a contradiction regarding the feature to which the orthogonality loss is applied. 
Paper's (Sec 3.2.2): The text states that the constraint is applied to the unified, post-fusion feature $\tilde{Z}$. The loss formula is as follows:
<img width="381" height="106" alt="Image" src="https://github.com/user-attachments/assets/f5ee022a-8301-43d9-8489-d0a436f2ad1f" />
Code Implementation: The code in `train_brats.py` and `ortho_loss.py` applies the loss to each of the individual, pre-fusion content features ($\mathcal{Z}_i$). Could you clarify which feature is the intended target for the orthogonality constraint? 

Any insight you could provide on these two points would be a great help to understanding the work more deeply. 

Thank you for your time and clarification!

**Comment by xionghonglin** [comment]
Thank you so much for your insightful questions and for taking the time to review our work so thoroughly. 

You are correct on both points. The issues you've identified in the train_brats.py file are due to my oversight. During the project, I used two separate repositories for multi-modal and single-modal input experiments. In an effort to unify the codebase for public release, I mistakenly merged some of the code from the single-modal version into the multi-modal training script.

To address your specific questions:
Q1. Content Alignment Loss
You are right. The content alignment loss should not be calculated with the zeroed-out features of missing modalities. The correct approach, as you suggested, is to compute this loss only among the content representations of the currently available modalities. This can be achieved by using the m_src variable in the training code to index the valid modalities.
Q2. Orthogonality Loss
Your observation here is also correct. The orthogonality loss is intended to be applied to the unified, post-fusion feature (Z), as described in the paper. The current implementation, which applies it to the individual pre-fusion features (Z_i ), is an error from the code merge. To fix this, the fused content features would need to be returned from models.lccd and used for the loss calculation.
Both of these corrections are relatively straightforward to implement. I apologize for the confusion this has caused.

Thank you once again for your diligence and for bringing these errors to my attention. I will work on correcting the public repository as soon as possible.

**Comment by xionghonglin** [comment]
Thank you so much for your insightful questions and for taking the time to review our work so thoroughly.

You are correct on both points. The issues you've identified in the train_brats.py file are due to my oversight. During the project, I used two separate repositories for multi-modal and single-modal input experiments. In an effort to unify the codebase for public release, I mistakenly merged some of the code from the single-modal version into the multi-modal training script.

To address your specific questions:
Q1. Content Alignment Loss
You are right. The content alignment loss should not be calculated with the zeroed-out features of missing modalities, which is an error from the code merge.  The correct approach is to compute this loss only among the content representations of the currently available modalities. This can be achieved by using the m_src variable in the training code to index the valid modalities.
Q2. Orthogonality Loss
Your observation here is also correct. The orthogonality loss is intended to be applied to the unified, post-fusion feature (Z), as described in the paper. The current implementation, which applies it to the individual pre-fusion features (Z_i ), is also an error from the code merge. To fix this, the fused content features would need to be returned from models.lccd and used for the loss calculation.
Both of these corrections are relatively straightforward to implement. I apologize for the confusion this has caused.

Thank you once again for your diligence and for bringing these errors to my attention. I will work on correcting the public repository as soon as possible.

**Comment by wenhaoqiao** [comment]
Hi author,

Thank you so much for the swift and incredibly clear response!

Your explanation regarding the repository merge clarifies everything perfectly. I'll look forward to the updated repository. Many thanks again for your help and for sharing this great project!

**Comment by m-jia1** [comment]
Hi author,
I have the same question as the previous questioner. Can I get your corrected code?  Thank you very much!
